src/pubsub/ua_pubsub_ns0.c

```diff
@@ -248,15 +266,26 @@ addPubSubConnectionConfig(UA_Server *server, UA_PubSubConnectionDataType *pubsub
     //connectionConfig.enabled = pubSubConnectionDataType.enabled;
     UA_Variant_setScalar(&connectionConfig.address, &networkAddressUrlDataType,
                          &UA_TYPES[UA_TYPES_NETWORKADDRESSURLDATATYPE]);
-    if(pubsubConnectionDataType->publisherId.type == &UA_TYPES[UA_TYPES_UINT32]){
-        connectionConfig.publisherId.numeric = * ((UA_UInt32 *) pubsubConnectionDataType->publisherId.data);
-    } else if(pubsubConnectionDataType->publisherId.type == &UA_TYPES[UA_TYPES_STRING]){
+
+
+    if (pubsubConnectionDataType->publisherId.type == &UA_TYPES[UA_TYPES_BYTE]) {
+        connectionConfig.publisherIdType = UA_PUBSUB_PUBLISHERID_BYTE;
+        connectionConfig.publisherId.publisherIdByte = *((UA_Byte *) pubsubConnectionDataType->publisherId.data);
+    } else if (pubsubConnectionDataType->publisherId.type == &UA_TYPES[UA_TYPES_UINT16]) {
+        connectionConfig.publisherIdType = UA_PUBSUB_PUBLISHERID_UINT16;
+        connectionConfig.publisherId.publisherIdUInt16 = *((UA_UInt16 *) pubsubConnectionDataType->publisherId.data);
+    } else if (pubsubConnectionDataType->publisherId.type == &UA_TYPES[UA_TYPES_UINT32]) {
+        connectionConfig.publisherIdType = UA_PUBSUB_PUBLISHERID_UINT32;
+        connectionConfig.publisherId.publisherIdUInt32 = *((UA_UInt32 *) pubsubConnectionDataType->publisherId.data);
+    } else if (pubsubConnectionDataType->publisherId.type == &UA_TYPES[UA_TYPES_UINT64]) {
+        connectionConfig.publisherIdType = UA_PUBSUB_PUBLISHERID_UINT64;
+        connectionConfig.publisherId.publisherIdUInt64 = *((UA_UInt64 *) pubsubConnectionDataType->publisherId.data);
+    } else if (pubsubConnectionDataType->publisherId.type == &UA_TYPES[UA_TYPES_STRING]) {
         connectionConfig.publisherIdType = UA_PUBSUB_PUBLISHERID_STRING;
-        UA_String_copy((UA_String *) pubsubConnectionDataType->publisherId.data, &connectionConfig.publisherId.string);
+        connectionConfig.publisherId.publisherIdString = *((UA_String *) pubsubConnectionDataType->publisherId.data);
```

Comment:
I don't understand how I can memcpy the entire union without if/else

Because input UA_PubSubConnectionDataType has publisherId of type Variant
how can I memcpy it and set publisherIdType and publisherId.byte etc. correctly?

or do you mean I should just set the enum type

and then memcpy the pubsubConnectionDataType->publisherId.data to connectionConfig.publisherId?

Reply:
thanks for the suggestion
I set the type within if/else and memcopied PublisherId union after that
Hope that is what you had in mind